server/internal/app/internal/middleware/assets_cors_middleware.go

```diff
@@ -0,0 +1,51 @@
+package middleware
+
+import (
+	"net/http"
+
+	"github.com/labstack/echo/v4"
+	"github.com/reearth/reearth/server/internal/usecase/gateway"
+	"github.com/reearth/reearthx/log"
+)
+
+func AssetsCORSMiddleware(domainChecker gateway.DomainChecker, allowedOrigins []string) func(echo.HandlerFunc) echo.HandlerFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			origin := c.Request().Header.Get("Origin")
+			if origin == "" {
+				return c.NoContent(http.StatusBadRequest)
+			}
+			allowedOrigin := ""
+			for _, allowed := range allowedOrigins {
+				if allowed == origin {
+					allowedOrigin = allowed
+					break
+				}
+			}
+			if allowedOrigin == "" {
+				domainResp, err := domainChecker.CheckDomain(c.Request().Context(), gateway.DomainCheckRequest{
+					Domain: origin,
+				})
+				if err != nil {
+					log.Errorfc(c.Request().Context(), "[AssetsCORSMiddleware] domain checker check domain err: %v", err)
+					return next(c)
+				}
+				if domainResp.Allowed {
+					c.Response().Header().Set("Access-Control-Allow-Origin", origin)
+				}
+			} else {
+				c.Response().Header().Set("Access-Control-Allow-Origin", allowedOrigin)
+			}
+
+			c.Response().Header().Set("Access-Control-Allow-Methods", "GET, OPTIONS")
+			c.Response().Header().Set("Access-Control-Allow-Headers", "*")
+			c.Response().Header().Set("Access-Control-Max-Age", "86400")
+
+			if c.Request().Method == "OPTIONS" {
+				return c.NoContent(http.StatusNoContent)
+			}
+
+			return next(c)
+		}
+	}
+}
```

Comment:
[memo]
main logic is here.
if we inject domain check gateway, we use to check wether origin is used as custom domain or not. If we don't inject domain check gateway, we check only allowed origins.